### PR TITLE
[Checkbox] Adds shape property

### LINKED ISSUE #70003: Proposal to add border radius attribute to CheckBox widget.
In most cases, users might want rounded corner radius of checkbox

### PR DESCRIPTION
**Pull request by yash1200**
## Description

Shape property added in Checkbox Widget.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/70003

## Tests

I added the following tests:

* Tests added in `test/material/checkbox_test.dart`

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by jinfagang** [comment]
How's this PR going?

**Comment by HansMuller** [comment]
If we were going to allow developers to tweak the shape of the checkbox "box" it would be with BorderRadiusGeometry, the property would be default null, and would be reflected in this WIP theme PR: https://github.com/flutter/flutter/pull/70311.

However if we're going to make the shape and appearance of checkboxes configurable, why not allow the developer to specify a BorderSide and a [shape](https://api.flutter.dev/flutter/painting/OutlinedBorder-class.html)?

**Comment by yash1200** [comment]
Then I think I should wait for #70311 to land first.

**Comment by jinfagang** [comment]
@yash1200 that PR already merged. When and where channel can we using this property as quick as possible?

**Comment by yash1200** [comment]
@jinfagang I have made the changes w.r.t #70311, I am waiting for @HansMuller to review these changes and if this PR merges, you can find these properties in `master` channel.

File: packages/flutter/lib/src/material/checkbox.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/checkbox.dart:244]: Please don't just reiterate the property name as documentation. Explain what the shape contributes to the checkbox visuals. 

If you refer to a significant type then use `[Foo]`.

If we're going to allow developers to specify the checkbox's shape, then the type of this property should be generic. In this case I'd suggest [OutlinedBorder](https://api.flutter.dev/flutter/painting/OutlinedBorder-class.html)

Please explain what this property's default value is in terms of this parameter, and the theme.

File: packages/flutter/lib/src/material/checkbox.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/checkbox.dart:247]: Please don't just reiterate the property name as documentation. Explain how this parameter affects the checkbox's visuals.

Please explain what this property's default value is in terms of this parameter, and the theme.

File: packages/flutter/lib/src/material/checkbox.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/checkbox.dart:405]: The default value of this property should be `Directionality.of(context) ?? TextDirection.ltr`

File: packages/flutter/lib/src/material/checkbox.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/checkbox.dart:562]: Can't assume that the shape is a rounded rectangle.

File: packages/flutter/lib/src/material/checkbox_theme.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/checkbox_theme.dart:139]: If we're going to support textDirection as a parameter, then we might as well lerp it here.

**Comment by skia-gold** [comment]
Gold has detected about 29 untriaged digest(s) on patchset 8.
View them at https://flutter-gold.skia.org/cl/github/70171

**Comment by skia-gold** [comment]
Gold has detected about 29 untriaged digest(s) on patchset 9.
View them at https://flutter-gold.skia.org/cl/github/70171

File: packages/flutter/lib/src/material/checkbox.dart
Comment (yash1200) [inline packages/flutter/lib/src/material/checkbox.dart:574]: With using this method, I'm facing this issue:
If shape is passed null, then a checkbox with black border, width = 2 and borderRadius = 1 is rendered.
If shape is defined as 
```
shape: RoundedRectangleBorder(
   borderRadius: BorderRadius.circular(5),
),
```
Then a checkbox with transparent border and borderRadius = 5 is rendered (This should be a checkbox with border color, width = 2 and borderRadius = 5)

I want to fallback to default side if it isn't provided
How should I proceed?

**Comment by skia-gold** [comment]
Gold has detected about 29 untriaged digest(s) on patchset 10.
View them at https://flutter-gold.skia.org/cl/github/70171

**Comment by skia-gold** [comment]
Gold has detected about 58 untriaged digest(s) on patchset 11.
View them at https://flutter-gold.skia.org/cl/github/70171

**Comment by skia-gold** [comment]
Gold has detected about 58 untriaged digest(s) on patchset 12.
View them at https://flutter-gold.skia.org/cl/github/70171

**Comment by skia-gold** [comment]
Gold has detected about 29 untriaged digest(s) on patchset 13.
View them at https://flutter-gold.skia.org/cl/github/70171

**Comment by skia-gold** [comment]
Gold has detected about 29 untriaged digest(s) on patchset 14.
View them at https://flutter-gold.skia.org/cl/github/70171

**Review by HansMuller — COMMENTED** [review]
This is getting really close. Sorry about the delays in getting a reviews.

**Comment by yash1200** [comment]
I think it would be confusing to have `side` as well as `shape.side`.

**Comment by HansMuller** [comment]
As I mentioned in https://github.com/flutter/flutter/pull/70171#discussion_r536451294, although just having a shape parameter is simpler, automatically substituting the default BorderSide when the shape's side is BorderSide.none prevents developers from creating a checkbox with no border. 

Adding a side parameter, like ButtonStyle.side, avoids that problem. And it's consistent with ButtonStyle.

**Comment by jinfagang** [comment]
@HansMuller when will this feature merge?

**Comment by HansMuller** [comment]
I hope it can be merged once this has been resolved: https://github.com/flutter/flutter/pull/70171#issuecomment-746924011

**Comment by yash1200** [comment]
@HansMuller please take a look at the changes :)

File: packages/flutter/lib/src/material/checkbox.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/checkbox.dart:644]: This method doesn't appear to draw anything.

File: packages/flutter/lib/src/material/checkbox.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/checkbox.dart:719]: It would help readability a little if `shape.copyWith(side: side).getOuterPath(outer)` was computed once and bound to a variable with a name like emptyCheckboxPath.

File: packages/flutter/lib/src/material/checkbox_theme.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/checkbox_theme.dart:139]: This shouldn't be commented out. Since BorderSide.lerp doesn't support null parameters (for now) you can define a private static utility method, and use ` _lerpSides(a?.side, b?.side, t)`

```dart
  // Special case because BorderSide.lerp() doesn't support null arguments
  static MaterialStateProperty<BorderSide?>? _lerpSides(MaterialStateProperty<BorderSide?>? a, MaterialStateProperty<BorderSide?>? b, double t) {
    if (a == null && b == null)
      return null;
    return _LerpSides(a, b, t);
  }
```

File: packages/flutter/test/material/checkbox_test.dart
Comment (HansMuller) [inline packages/flutter/test/material/checkbox_test.dart:924]: Should verify that we're actually painting the right thing with expect(... paints ...) as elsewhere in this file. The current test only proves that the shape and side are bound to instance vars.

File: packages/flutter/test/material/checkbox_list_tile_test.dart
Comment (HansMuller) [inline packages/flutter/test/material/checkbox_list_tile_test.dart:53]: This kind of thing is typically formatted (elsewhere in this file) as 
```dart
expect(getCheckboxListTileRenderer(), 
  paints
    ..path(color: const Color(0xff1e88e5))
    ..path(color: const Color(0xFFFFFFFF))
 );
```

It would help if the colors were bound to vars with names that indicated what part of the checkbox the color was being applied to.

File: packages/flutter/test/material/checkbox_test.dart
Comment (HansMuller) [inline packages/flutter/test/material/checkbox_test.dart:433]: Can be on one line (here and elsewehere)

File: packages/flutter/lib/src/material/checkbox.dart
Comment (yash1200) [reply packages/flutter/lib/src/material/checkbox.dart:644]: This function is drawing the outer path of the checkbox.

Comment (yash1200) [reply packages/flutter/lib/src/material/checkbox.dart:644]: And all the other required changes are done :)

**Review by HansMuller — APPROVED** [review]
LGTM

Thanks for sticking with this!

**Comment by jinfagang** [comment]
@yash1200 Would u help to merge this PR?

**Comment by yash1200** [comment]
@jinfagang I can't do anything, we will have to wait for `Google testing` test to pass to get it merged automatically.

**Comment by jinfagang** [comment]
@yash1200  So it can be used on master now? Any documentation?

**Comment by yash1200** [comment]
@jinfagang  Yes it can be used on the latest master channel, you can give the shape to checkbox using `shape` parameter and the side of the checkbox can be configured with `side` property.

**Comment by jinfagang** [comment]
@yash1200  Nice nice. can I remove the border and set only backgronud color with a border radius?

**Comment by netgfx** [comment]
Can't find it in the documentation though https://api.flutter.dev/flutter/material/Checkbox-class.html

**Comment by yash1200** [comment]
Oh, I'll make a PR for that. Thanks for letting me know.

**Pull request merged**
